UNL-CSE-AMBASSADORS/VITA
# 83 appointment scheduling rules

This PR removes logic about appointment times from the front-end and moves it all to the SQL queries and PHP. This also helped reduce the number of calls to the server from the client.

The database was updated to include an AppointmentTime table. This table is placed between Site and Appointment on the ERD. It takes holds individual datetimes for sites with the assumption that there is a Shift that covers the time frame of the AppointmentTime. By separating the times from the individual appointments, there will no longer be times placed in the appointment table that come from the front-end, i.e. all time stamps are set ahead of time, appointments just reference them. This also makes it easier to store logic surrounding the number of appointments that can be made. There are two columns that help with that:
* ```maximumNumberOfAppointments``` allows there to be a set number of appointments that can be scheduled. If not set, the number of appointments that can be made at a certain date/site/time can be calculated based on the number of volunteers that are signed up.
* ```percentageAppointments``` is used to limit the number of appointments that are made by the system, leaving spots for walk-in clients.

The test data was also updated to use predefined AppointmentTimes. In the test data Site 1 does not have a maximum, so appointment availability is based on the number of volunteers that have signed up. Site 2 has a maximum of 30 set. That being said, with 200 random appointments being made, many of the date/site/time options are full. Site 2 usually does still have availability.

**TESTING**

* Run the updated SQL scripts.
* Ensure that the API script works by visiting [localhost/server/api/appointments/getTimes.php](http://localhost/server/api/appointments/getTimes.php) (I recommend getting the Chrome Extension, JSONView to make reading this easier.)
* Go to the /signup page.
* Ensure that the date picker, site picker, and time picker all still work.
* Ensure that submitting the data still works and that it is stored correctly in the database.

**ERD**
Here's the ERD after these changes:
![databaseerd](https://user-images.githubusercontent.com/9295744/34136722-f5ccdb08-e42c-11e7-9a56-b6fe8357d6d5.png)
